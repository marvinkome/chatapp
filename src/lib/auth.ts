export async function auth(endpoint: string, body: any) {
    const resp = await fetch(`http://localhost:5000/auth/${endpoint}`, {
        method: 'POST',
        body: JSON.stringify(body),
        headers: {
            'Content-Type': 'application/json'
        }
    });

    return resp.json();
}
