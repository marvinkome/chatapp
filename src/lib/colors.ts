export default {
    text_light: 'hsl(0, 0%, 45%)',
    text_title: 'hsl(0, 0%, 20%)',
    primary: '#42ED86'
};
