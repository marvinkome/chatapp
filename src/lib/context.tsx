import * as React from 'react';

export function contextConnect(Context: React.Context<{}>, mapProvider: (context: any) => object) {
    return (Component: any) => {
        return class WrappedComponent extends React.Component<any> {
            static displayName = `withContext(${Component.displayName})`;

            render() {
                return (
                    <Context.Consumer>
                        {(context: any) => {
                            const props = mapProvider(context);
                            return <Component {...this.props} {...props} />;
                        }}
                    </Context.Consumer>
                );
            }
        };
    };
}
