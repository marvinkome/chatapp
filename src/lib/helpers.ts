import moment from 'moment';
import { AsyncStorage } from 'react-native';

export function formatDate(date: any, relative?: boolean) {
    const mDate = moment(date);
    return relative ? mDate.fromNow() : mDate.format('h:mm A');
}

export const clientId = async () => {
    const token = await AsyncStorage.getItem('client_id');
    return token;
};
