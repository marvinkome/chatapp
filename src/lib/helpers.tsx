import * as React from 'react';
import moment, { Moment } from 'moment';
import { AsyncStorage } from 'react-native';
import { MessageBarManager } from 'react-native-message-bar';

export function formatDate(date: any, full?: boolean) {
    const currDate = moment();
    const myDate = moment(date);
    const today = currDate.clone().startOf('day');
    const yday = currDate
        .clone()
        .subtract(1, 'days')
        .startOf('day');

    function isToday(mdate: Moment) {
        return mdate.isSame(today, 'd');
    }

    function isYesterday(mdate: Moment) {
        return mdate.isSame(yday, 'd');
    }

    if (isToday(myDate)) {
        return full ? myDate.format('[Today at] LT') : myDate.format('h:mm A');
    }

    if (isYesterday(myDate)) {
        return full ? myDate.format('[Yesterday at] LT') : 'Yesterday';
    }

    return full ? myDate.format('D MMM YYYY LT') : myDate.format('D[/]M[/]YY');
}

export async function clientId() {
    const token = await AsyncStorage.getItem('client_id');
    return token;
}

export function showAlert(msg: string, type?: string, options?: any) {
    MessageBarManager.showAlert({
        ...options,
        message: msg,
        shouldHideAfterDelay: false,
        alertType: type || 'warning',
        messageNumberOfLines: 1,
        durationToShow: 0,
        durationToHide: 0,
        messageStyle: {
            fontFamily: 'PT_Sans',
            textAlign: 'center',
            fontSize: 14
        }
    });
}

export function hideAlert() {
    MessageBarManager.hideAlert();
}

export function networkErrHandler(
    client: any,
    query: any,
    Comp: React.ComponentClass
) {
    // get data from the cache
    const cache = client.readQuery({ query });

    // show error
    // TODO: Change error message
    showAlert("You're offline", 'error', {
        shouldHideOnTap: false
    });

    // @ts-ignore
    return cache.user ? <Comp data={cache} /> : null;
}
