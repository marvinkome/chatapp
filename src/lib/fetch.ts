import { AsyncStorage } from 'react-native';
import { auth } from './auth';

const clientId = async () => {
    const token = await AsyncStorage.getItem('client_id');
    return token;
};

async function customFetch(uri: string, options: any) {
    const token = await clientId();

    // if there's token continue
    if (token) {
        // authenticate and get an access token
        try {
            const data = await auth('', { client_token: token });
            const accessToken = data.payload;
            options = {
                header: {
                    authorization: `Bearer ${accessToken}`
                }
            };
        } catch {
            // what do we do?
            // log him out? yeah
            // logout
        }
    }

    return fetch(uri, options);
}

export default customFetch;
