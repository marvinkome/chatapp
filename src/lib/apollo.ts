import ApolloClient from 'apollo-client';
import { AsyncStorage } from 'react-native';
import { split } from 'apollo-link';
import { WebSocketLink } from 'apollo-link-ws';
import { createHttpLink } from 'apollo-link-http';
import { setContext } from 'apollo-link-context';
import { getMainDefinition } from 'apollo-utilities';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { CachePersistor } from 'apollo-cache-persist';
import { clientId } from './helpers';
import { DefaultOptions } from 'apollo-client/ApolloClient';

const defaultOptions: DefaultOptions = {
    watchQuery: {
        fetchPolicy: 'cache-and-network',
        errorPolicy: 'all'
    },
    query: {
        fetchPolicy: 'cache-and-network',
        errorPolicy: 'all'
    },
    mutate: {
        errorPolicy: 'all'
    }
};

async function Link() {
    const token = await clientId();

    // http link
    const httpLink = createHttpLink({
        uri: 'http://localhost:3000/graphql'
    });

    // auth link
    const authLink = setContext((_, { headers }) => {
        return {
            headers: {
                ...headers,
                authorization: token ? `Bearer ${token}` : ''
            }
        };
    }).concat(httpLink);

    // ws link
    const wsLink = new WebSocketLink({
        uri: 'ws://localhost:3000/graphql',
        options: {
            reconnect: true,
            connectionParams: {
                authToken: token
            }
        }
    });

    // filter link
    const link = split(
        ({ query }) => {
            // @ts-ignore
            const { kind, operation } = getMainDefinition(query);
            return (
                kind === 'OperationDefinition' && operation === 'subscription'
            );
        },
        wsLink,
        authLink
    );

    return link;
}

export default async () => {
    const cache = new InMemoryCache({
        cacheRedirects: {
            User: {
                group: (_, args, { getCacheKey }) =>
                    getCacheKey({ __typename: 'Group', id: args.groupId })
            }
        }
    });

    const persistor = new CachePersistor({
        cache,
        storage: AsyncStorage
    });

    const link = await Link();

    const client = new ApolloClient({
        link,
        cache,
        defaultOptions
    });

    return {
        cache,
        client,
        persistor
    };
};
