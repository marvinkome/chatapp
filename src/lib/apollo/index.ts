import ApolloClient from 'apollo-client';
import { DefaultOptions } from 'apollo-client/ApolloClient';
import { AsyncStorage, NetInfo } from 'react-native';
import introspectionQueryResultData from './fragmentTypes.json';

// links
import Link from './link';

import { SyncOfflineMutation } from './syncMutation';

import {
    InMemoryCache,
    IntrospectionFragmentMatcher
} from 'apollo-cache-inmemory';
import { CachePersistor } from 'apollo-cache-persist';
import { clientId } from '../helpers';

const defaultOptions: DefaultOptions = {
    watchQuery: {
        fetchPolicy: 'cache-first',
        errorPolicy: 'all'
    },
    query: {
        fetchPolicy: 'cache-first',
        errorPolicy: 'all'
    },
    mutate: {
        errorPolicy: 'all'
    }
};

export default async () => {
    const token = await clientId();

    const fragmentMatcher = new IntrospectionFragmentMatcher({
        introspectionQueryResultData
    });

    const cache = new InMemoryCache({
        cacheRedirects: {
            User: {
                group: (_, args, { getCacheKey }) =>
                    getCacheKey({ __typename: 'Group', id: args.groupId })
            }
        },
        fragmentMatcher
    });

    const persistor = new CachePersistor({
        cache,
        storage: AsyncStorage,
        trigger: 'write'
    });

    const { link, queueLink, mutationLink } = Link(token, cache);

    const client = new ApolloClient({
        link,
        cache,
        defaultOptions
    });

    // add listeners to queue operations
    // const isOnline = (e: any) => e !== 'none' && e !== 'unknown';
    // NetInfo.addEventListener('connectionChange', (e) => {
    //     // @ts-ignore
    //     if (isOnline(e.type)) {
    //         mutationLink.open(client);
    //         queueLink.open();
    //     } else {
    //         mutationLink.close();
    //         queueLink.close();
    //     }
    // });

    // const syncMutation = new SyncOfflineMutation(client, AsyncStorage);
    // await syncMutation.init();
    // await syncMutation.sync();
    // await syncMutation.clearOfflineData();

    return {
        cache,
        client,
        persistor
    };
};
