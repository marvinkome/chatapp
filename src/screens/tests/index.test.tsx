import * as React from 'react';
import { create } from 'react-test-renderer';
import App from '../index';
import { flushPromises } from '../../lib/tests';

jest.unmock('react-navigation');
describe('screen tests', () => {
    const comp = create(<App />);
    flushPromises();

    it('renders correctly, <snapshot>', () => {
        expect(comp.toJSON()).toMatchSnapshot();
    });
});
