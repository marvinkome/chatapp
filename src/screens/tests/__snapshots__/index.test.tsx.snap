// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`screen tests renders correctly, <snapshot> 1`] = `null`;
